src/H5C.c

```diff
@@ -416,7 +416,7 @@ H5C_prep_for_file_close(H5F_t *f)
 
 #ifdef H5_HAVE_PARALLEL
     if ((H5F_INTENT(f) & H5F_ACC_RDWR) && !image_generated && cache_ptr->aux_ptr != NULL &&
-        f->shared->fs_persist) {
+        f->shared->sblock && f->shared->fs_persist) {
```

Comment:
Don't try to serialize the cache if we failed to init the superblock, because `H5MF_settle_raw_data_fsm` assumes it's available and would crash otherwise. Also, for self-referential free space allocators, `H5MF_settle_meta_data_fsm` causes them to allocate space for themselves, which later leads to a failure in the assertion `H5AC_cache_is_clean(f, H5AC_RING_MDFSM)` in `H5F__dest`.